File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        TreeNode left;
        TreeNode right;
        if (p.val < q.val) {
            left = p;
            right = q;
        } else {
            left = q;
            right = p;
        }
        if (left.val < root.val && root.val < right.val) return root;
        if (left.val == root.val) return left;
        if (right.val == root.val) return right;
        if (root.val < left.val) return lowestCommonAncestor(root.right, left, right);
        return lowestCommonAncestor(root.left, left, right);
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.py
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        nodeDiameter = {}

        def levelTree(node: Optional[TreeNode], nodeDiameter: dict) -> int:
            if node is None:
                return 0

            leftLevel = levelTree(node.left, nodeDiameter)
            rightLevel = levelTree(node.right, nodeDiameter)

            nodeDiameter[node.val] = leftLevel + rightLevel

            return 1 + max(leftLevel, rightLevel)

        if root is None:
            return 0

        rootLevel = levelTree(root.left, nodeDiameter) + \
                    levelTree(root.right, nodeDiameter)

        nodeDiameter[root.val] = rootLevel
        print(nodeDiameter)
        mx = 0
        for v in nodeDiameter.values():
            mx = max(mx, v)

        return mx